doc/adding_lints.md

```diff
@@ -0,0 +1,450 @@
+## Adding a new lint
+
+You are probably here because you want to add a new lint to Clippy. If this is
+the first time you're contributing to Clippy, this document guides you through
+creating an example lint from scratch.
+
+To get started, we will create a lint that detects functions called `foo`,
+because that's clearly a non-descriptive name.
+
+* [Setup](#Setup)
+* [Testing](#Testing)
+* [Rustfix tests](#Rustfix-tests)
+* [Lint declaration](#Lint-declaration)
+* [Lint passes](#Lint-passes)
+* [Emitting a lint](#Emitting-a-lint)
+* [Adding the lint logic](#Adding-the-lint-logic)
+* [Documentation](#Documentation)
+* [Running rustfmt](#Running-rustfmt)
+* [Debugging](#Debugging)
+* [PR Checklist](#PR-Checklist)
+* [Cheatsheet](#Cheatsheet)
+
+### Setup
+
+Clippy depends on the current git master version of rustc, which can change rapidly. Make sure you're
+working near rust-clippy's master, and use the `setup-toolchain.sh` script to configure the appropriate
+toolchain for this directory.
+
+### Testing
+
+Let's write some tests first that we can execute while we iterate on our lint.
+
+Clippy uses UI tests for testing. UI tests check that the output of Clippy is
+exactly as expected. Each test is just a plain Rust file that contains the code
+we want to check. The output of Clippy is compared against a `.stderr` file.
```

Comment:
"Note that you don't have to create this yourself, we'll get to generating `.stderr` files further down"

Reply:
Added in https://github.com/rust-lang/rust-clippy/pull/3824/commits/e9d37ca79a7e457a664a032e736be2eb9b9af3c7